ddtrace/tracer/remote_config.go

```diff
@@ -23,23 +25,102 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/remoteconfig/state"
 )
 
+// configData represents one config file received from Remote Config.
 type configData struct {
-	Action        string    `json:"action"`
-	ServiceTarget target    `json:"service_target"`
-	LibConfig     libConfig `json:"lib_config"`
+	Action        string      `json:"action"`
+	ServiceTarget target      `json:"service_target"`
+	K8sTargetV2   k8sTargetV2 `json:"k8s_target_v2"`
+	LibConfig     libConfig   `json:"lib_config"`
 }
 
 type target struct {
 	Service string `json:"service"`
 	Env     string `json:"env"`
 }
 
+type k8sTargetV2 struct {
+	ClusterTargets []clusterTarget `json:"cluster_targets"`
+}
+
+type clusterTarget struct {
+	ClusterName       string   `json:"cluster_name"`
+	Enabled           bool     `json:"enabled"`
+	EnabledNamespaces []string `json:"enabled_namespaces"`
+}
+
+// priority returns the order in which this config should be applied, relative
+// to other configs. The more specific the config's targeting of the current
+// process is, the higher the priority; configs with higher priority override
+// configs with lower priority.
+func (c configData) priority() int {
+	isSingleEnvironment := c.ServiceTarget.Env != "" && c.ServiceTarget.Env != "*"
+	isSingleService := c.ServiceTarget.Service != "" && c.ServiceTarget.Service != "*"
+	isClusterTarget := len(c.K8sTargetV2.ClusterTargets) > 0
+
+	switch {
+	case isSingleEnvironment && isSingleService:
+		return 5
+	case isSingleService:
+		return 4
+	case isSingleEnvironment:
+		return 3
+	case isClusterTarget:
+		return 2
+	default:
+		return 1
+	}
+}
+
+// mergeConfigsByPriority sorts the configs by priority (i.e. targeting
+// specificity) and merges them into a single libConfig. For each field, the
+// value from the highest-priority config with a non-nil value for that field is
+// used.
+func mergeConfigsByPriority(configs []configData) libConfig {
+	if len(configs) == 0 {
+		return libConfig{}
+	}
+
+	sort.Slice(configs, func(i, j int) bool {
+		return configs[i].priority() < configs[j].priority()
+	})
```

Comment:
Is it possible that two items have the same priority? What happens in that case? If the former is possible, I think this is behaves as an unstable sort, and it could lead to nondeterministic overrides.

Reply:
The intention is to not have multiple configs at the same level. The other tracers also do unstable sorts, so I thought I'd make it unstable to not promise anything. But I've changed it now to a stable sort to simplify the conversation.